### ISSUE TITLE
`fetch` result is cached between dev mode runs

### ISSUE THREAD

**Issue by joulev** [vercel/next.js#55702]
### Link to the code that reproduces this issue

https://github.com/joulev/debug/tree/nextjs-fetch-result-cached-between-dev-mode-runs

### To Reproduce

1. `next dev`
2. A random string *x* should be printed

    ```
     ✓ Compiled /page in 967ms (395 modules)
    8079aa87a04126b2
    ```

4. Ctrl+C
5. `next dev` again
6. A random string *y* should be printed

    ```
     ✓ Compiled /page in 798ms (395 modules)
    8079aa87a04126b2
    ```

8. Ctrl+C
9. `rm -rf .next`
10. `next dev` again
11. A random string *z* should be printed

    ```
     ✓ Compiled /page in 955ms (395 modules)
    6a876542c70550cd
    ```

### Current vs. Expected behavior

Current behaviour: *x* = *y* ≠ *z*
Expected behaviour: *x* ≠ *y* ≠ *z*

because dev mode should only perform request memorisation and shouldn't cache result across different page renders or dev mode runs. At least that's how dev mode has been working for me.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Thu Aug 17 21:23:02 PDT 2023; root:xnu-10002.1.11~3/RELEASE_ARM64_T8112
Binaries:
  Node: 20.3.1
  npm: 9.6.7
  Yarn: 1.22.19
  pnpm: 8.6.12
Relevant Packages:
  next: 13.5.2
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

* `export const dynamic = "force-dynamic"`: the bug is **not** reproduced
* `export const revalidate = 0`: the bug is **not** reproduced
* `export const runtime = "edge"`: the bug is reproduced (although supposedly the edge runtime makes the route dynamic at the moment)
* Using `cookies()` inside the page: the bug is reproduced (although the page should become dynamic implicitly)

---

This behaviour is similar to `unstable_cache`: the following also persists the result cross dev runs:

```tsx
import { unstable_cache } from "next/cache";

const getValue = unstable_cache(async () => new Date().toString());

export default async function Page() {
  const value = await getValue();
  return <div>{value}</div>;
}
```

---

I will investigate further and update this post when I find the canary version that caused this bug, but at least the bug was not here in v13.4.20-canary.18 and is reproduced in v13.5.1.

**Comment by joulev** [comment]
Seems like this "bug" has been here for quite a few stable versions. Considering `unstable_cache` also behaves like that, I'm wondering if this is even a bug... there is a good chance this is actually intended.

**Comment by icyJoseph** [comment]
You can blast the `.next` folder and that takes care of it right?

A while back, I was playing around with interceptors to make a server side UI to see, RSC requests, and found this too.

**Comment by joulev** [comment]
Yup, blasting the `.next` folder is the solution to this.

I'm fine with this being the intended behaviour (after all this is the behaviour for prod mode), but it should work consistently for implicit dynamic rendering too, not just when I explicitly say `revalidate = 0` or `dynamic = "force-dynamic"`.

**Comment by icyJoseph** [comment]
Yeah, but you know what other quirk it has, if you run next dev, and use API routes as you render. The typical, `fetch("https://localhost:3000/some/api/route/on-the-same-server")`, and then run a build, it works...

So people learning Next.js are convinced everything is fine, so they push, and it starts to fail on the deploy environment...

**Comment by icyJoseph** [comment]
> Yeah, we on the Next.js Discord server even went as far as writing a [long essay](https://nextjs-faq.com/fetch-api-in-rsc) on why fetching server endpoints in server components or `getServerSideProps` or similar places is an anti-pattern and highly recommended against.

Nice, I find people doing that, over and over on the discussion threads, I've even to just start to link back old answers and such, but it is a very extended anti-pattern.

So far I had been using this https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props#getserversideprops-or-api-routes as source of truht, I'll take a look at that FAQ link, maybe it'll save some time to just forward people there.

However, unfortunately, this anti-pattern is somehow validated during learning, when a person does `next dev` -> `next build` and it ~seems to succeed~ works locally, but not on the deploy environment.

**Comment by joulev** [comment]
Yeah... to be honest, I think `next dev` and `next build` should not depending on existing content of `.next`. Different `.next` before running them shouldn't yield different outputs like this. This might somehow be intended by the team but not the behaviour a typical user would expect.

**Comment by SachaG** [comment]
I just ran into a similar issue. 

I was calling `await fetchFoo()` at three different points from within a RSC, and noticed that the first two calls would give one result (in this case an older version of `foo` where `foo.bar` doesn't exist), and the third one another result (a new version where `foo.bar` does exist). 

Sure enough clearing `.next` and restarting fixed it (all three calls now return the correct, newer version). 

I don't think this can possibly be the intended behavior? At the very least different calls within the same request should return the same result?
